addons/sale_coupon/models/coupon_program.py

```diff
@@ -147,14 +147,13 @@ def _filter_not_ordered_reward_programs(self, order):
         Returns the programs when the reward is actually in the order lines
         """
         programs = self.env['coupon.program']
+        order_products = order.order_line.product_id
         for program in self:
-            if program.reward_type == 'product' and \
-               not order.order_line.filtered(lambda line: line.product_id == program.reward_product_id):
+            if program.reward_type == 'product' and program.reward_product_id not in order_products:
                 continue
-            elif program.reward_type == 'discount' and program.discount_apply_on == 'specific_products' and \
-               not order.order_line.filtered(lambda line: line.product_id in program.discount_specific_product_ids):
+            elif program.reward_type == 'discount' and program.discount_apply_on == 'specific_products' and not any(dp in order_products for dp in program.discount_specific_product_ids):
```

Comment:
`not any(dp in order_products for dp in program.discount_specific_product_ids)` --> `not program.discount_specific_product_ids & order_products`
no ?

Reply:
I thought about it, but the intersect would be slightly less performant, no?
any + in both short circuit as soon as a match is found

Reply:
Please keep the line split, it's a nightmare to have too long lines in an editor/github :D.
Also, please rename `dp` to `product` in the lambda, it's clearer (when I read dp, I think of decimal precision, not product ^^).

Reply:
Done ;)